crates/oxc_parser/src/lexer/simd.rs

```diff
@@ -0,0 +1,86 @@
+//! Lexer methods using portable-SIMD
+//! See:
+//!   * <https://github.com/rust-lang/portable-simd/blob/master/beginners-guide.md>
+//!   * <https://rapidjson.org/md_doc_internals.html#SkipwhitespaceWithSIMD>
+//!   * <https://lemire.me/blog/2017/01/20/how-quickly-can-you-remove-spaces-from-a-string>
+
+use std::simd::{Simd, SimdPartialEq, ToBitMask};
+
+const ELEMENTS: usize = 16;
+type SimdVec = Simd<u8, ELEMENTS>;
+
+#[derive(Debug)]
+pub struct SkipWhitespace {
+    /// Total offset
+    pub offset: usize,
+
+    /// Found multiline comment end '*/'?
+    pub found: bool,
+
+    /// Found newline inside the comment?
+    pub newline: bool,
+
+    lf: SimdVec,
+    cr: SimdVec,
+    space: SimdVec,
+    tab: SimdVec,
+}
+
+impl SkipWhitespace {
+    pub fn new(newline: bool) -> Self {
+        Self {
+            offset: 0,
+            found: false,
+            newline,
+            lf: SimdVec::splat(b'\n'),
+            cr: SimdVec::splat(b'\r'),
+            space: SimdVec::splat(b' '),
+            tab: SimdVec::splat(b'\t'),
+        }
+    }
+
+    pub fn simd(mut self, bytes: &[u8]) -> Self {
+        let (chunks, remainder) = bytes.as_chunks::<ELEMENTS>();
+
+        for chunk in chunks {
+            self.check_chunk(chunk);
+            if self.found {
+                return self;
+            }
+        }
+
+        if !remainder.is_empty() {
+            // Align the last chunk for avoiding the use of a scalar version
```

Comment:
If you ensure that the input data is padded (e.g. suffixed with 64 0x00 bytes), then you don't need this special case.

Reply:
`remainder` isn't aligned, it's a slice of the `bytes` input.

Reply:
If the input is 20 bytes, then the first 16 bytes are handled in the `for chunk in chunks` loop, and the remaining 4 bytes are handled by this code (which adds 12 zero bytes at the end). I assume this is what you mean when you say "`remainder` isn't aligned".

You can guarantee that `remainder` is always empty or unimportant (thus you don't need to check for it) if you add enough zeros at the end of `bytes`. I do this in quick-lint-js when I load the file from the filesystem, so everything can safely assume that the extra bytes are present. (I have a dedicated type for this called `padded_string_view` which doesn't allow breaking this invariant.)

Reply:
This is clever! I'll try it out.